**Цель работы:** осуществить предварительную обработку данных CSV-файла (visits.csv), выявить и устранить проблемы в этих данных.

Набор данных visits.csv содержит сведения о пользовательских сессиях в интернет-магазине.  
Каждая строка представляет собой визит пользователя со следующими признаками:
1. user_id — уникальный идентификатор пользователя; 
2. region — страна / регион пользователя;
3. device — тип устройства (iPhone, Mac, Android, PC);
4. channel — канал привлечения (organic, TipTop, FaceBoom, MediaTornado);
5. session_start — дата и время начала сессии;
6. session_end — дата и время завершения сессии;
7. time_session — длительность сессии в минутах;
8. click_count — число переходов / кликов за сессию;
9. buy_count — количество оформленных покупок;
10. price — общая стоимость покупок за сессию;
11. age — возраст пользователя.

# 1. Загрузка набора данных

In [14]:
import pandas as pd

df = pd.read_csv('visits.csv', sep=';')
print(f"Размер датасета: {df.shape[0]} строк, {df.shape[1]} столбцов")

Размер датасета: 954 строк, 11 столбцов


### 1.1 Вывод первых 20 и последних 10 строк набора данных

In [15]:
df.head(20)

,user_id,region,device,channel,session_start,session_end,time_session,click_count,buy_count,price,age
0,278965908054,United States,iPhone,organic,01.05.2019 4:46,01.05.2019 4:47,1.0,0.0,0.0,0.0,31
1,244878945432,United States,Mac,organic,01.05.2019 16:44,01.05.2019 16:45,1.0,0.0,0.0,0.0,40
2,590706206550,United States,Mac,organic,01.05.2019 14:09,01.05.2019 15:32,83.0,28.0,6.0,14570.0,50
3,326433527971,United States,Android,TipTop,01.05.2019 0:29,01.05.2019 0:54,25.0,8.0,3.0,4411.0,38
4,349773784594,United States,Mac,organic,01.05.2019 3:33,01.05.2019 3:57,24.0,8.0,3.0,4411.0,63
5,90242400005,United States,Android,organic,01.05.2019 9:03,01.05.2019 10:08,65.0,22.0,4.0,11310.0,62
6,43958116050,United States,iPhone,organic,01.05.2019 9:37,01.05.2019 10:00,23.0,8.0,3.0,4411.0,60
7,15174034813,United States,Mac,organic,01.05.2019 13:33,01.05.2019 13:35,2.0,0.0,0.0,0.0,12
8,446013509831,United States,iPhone,organic,01.05.2019 14:44,01.05.2019 15:41,57.0,19.0,4.0,9872.0,52
9,100970711362,United States,Mac,FaceBoom,01.05.2019 6:20,01.05.2019 6:54,34.0,12.0,3.0,6326.0,25


In [16]:
df.tail(10)

,user_id,region,device,channel,session_start,session_end,time_session,click_count,buy_count,price,age
944,159471803772,United States,iPhone,organic,04.05.2019 0:24,04.05.2019 0:58,34.0,12.0,3.0,6326.0,47
945,917992220173,United States,Android,organic,03.05.2019 23:17,03.05.2019 23:37,20.0,7.0,2.0,3739.0,46
946,977891886135,United States,Android,organic,03.05.2019 3:26,03.05.2019 3:46,20.0,7.0,2.0,3739.0,56
947,302414975073,United States,Android,TipTop,04.05.2019 5:26,04.05.2019 6:19,53.0,18.0,4.0,9394.0,20
948,579579190220,United States,Android,TipTop,04.05.2019 17:42,04.05.2019 18:06,24.0,8.0,3.0,4411.0,62
949,405931746536,United States,Mac,organic,03.05.2019 8:51,03.05.2019 9:11,20.0,7.0,2.0,3739.0,34
950,856975653366,United States,iPhone,organic,03.05.2019 4:24,03.05.2019 4:45,21.0,7.0,2.0,3739.0,22
951,510919926079,United States,Mac,organic,03.05.2019 2:47,03.05.2019 3:07,20.0,7.0,2.0,3739.0,41
952,531513927329,United States,Mac,organic,04.05.2019 13:48,04.05.2019 14:08,20.0,7.0,2.0,3739.0,53
953,136097633625,United States,Android,TipTop,04.05.2019 16:42,04.05.2019 17:01,19.0,7.0,2.0,3739.0,30


# 2. Первичный обзор и оценка качества данных

### Обзор данных и оценка структуры методом info()

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 954 entries, 0 to 953
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   user_id        954 non-null    int64  
 1   region         953 non-null    str    
 2   device         953 non-null    str    
 3   channel        954 non-null    str    
 4   session_start  954 non-null    str    
 5   session_end    954 non-null    str    
 6   time_session   954 non-null    float64
 7   click_count    954 non-null    float64
 8   buy_count      954 non-null    float64
 9   price          954 non-null    float64
 10  age            954 non-null    int64  
dtypes: float64(4), int64(2), str(5)
memory usage: 82.1 KB


### Оценка числовых признаков методом describe()

In [18]:
df.describe()

,user_id,time_session,click_count,buy_count,price,age
count,9.540000e+02,954.000000,954.000000,954.000000,954.000000,954.000000
mean,4.951807e+11,29.062893,9.830189,2.527254,5341.559748,39.810273
std,2.899488e+11,27.134816,9.209039,1.602003,5060.247685,13.038029
min,2.346325e+09,0.000000,0.000000,0.000000,0.000000,10.000000
25%,2.295122e+11,9.000000,3.000000,2.000000,2301.000000,29.000000
50%,4.984267e+11,21.000000,7.000000,2.000000,3739.000000,39.000000
75%,7.525332e+11,42.000000,14.000000,3.000000,7284.000000,50.000000
max,9.985130e+11,262.000000,88.000000,15.000000,56546.000000,71.000000


#### Проверка и нормализация наименований столбцов

In [19]:
print("Исходные названия столбцов:", list(df.columns))
df.columns = df.columns.str.strip().str.lower()
print("Нормализованные названия столбцов:", list(df.columns))

Исходные названия столбцов: ['user_id ', 'region', 'device', 'channel', 'session_start', 'session_end', 'time_session', 'click_count', 'buy_count', 'price', 'age']
Нормализованные названия столбцов: ['user_id', 'region', 'device', 'channel', 'session_start', 'session_end', 'time_session', 'click_count', 'buy_count', 'price', 'age']


## 3. Проверка и устранение пропущенных значений

In [20]:
print("Количество пропусков до обработки:")
print(df.isna().sum())

mode_region = df['region'].mode()[0]
mode_device = df['device'].mode()[0]

df['region'] = df['region'].fillna(mode_region)
df['device'] = df['device'].fillna(mode_device)

print("\nКоличество пропусков после обработки:")
print(df.isna().sum())
print("\nИтого оставшихся пропусков:", df.isna().sum().sum())

Количество пропусков до обработки:
user_id          0
region           1
device           1
channel          0
session_start    0
session_end      0
time_session     0
click_count      0
buy_count        0
price            0
age              0
dtype: int64

Количество пропусков после обработки:
user_id          0
region           0
device           0
channel          0
session_start    0
session_end      0
time_session     0
click_count      0
buy_count        0
price            0
age              0
dtype: int64

Итого оставшихся пропусков: 0


## 4. Проверка и устранение дубликатов

In [21]:
print("Количество полных явных дубликатов:", df.duplicated().sum())

print("\nУникальные значения region до унификации:")
print(df['region'].value_counts())

print("\nУникальные значения device до унификации:")
print(df['device'].value_counts())

df['region'] = df['region'].replace({'USA': 'United States'})
df['device'] = df['device'].replace({'MAC': 'Mac', 'IPHONE': 'iPhone'})

print("\nРаспределение region после унификации:")
print(df['region'].value_counts())

print("\nРаспределение device после унификации:")
print(df['device'].value_counts())

Количество полных явных дубликатов: 0

Уникальные значения region до унификации:
region
United States    901
Russia            52
USA                1
Name: count, dtype: int64

Уникальные значения device до унификации:
device
iPhone     420
Mac        243
Android    185
PC         103
IPHONE       2
MAC          1
Name: count, dtype: int64

Распределение region после унификации:
region
United States    902
Russia            52
Name: count, dtype: int64

Распределение device после унификации:
device
iPhone     422
Mac        244
Android    185
PC         103
Name: count, dtype: int64


## 5. Проверка и преобразование типов данных

In [22]:
df['session_start'] = pd.to_datetime(df['session_start'], format='%d.%m.%Y %H:%M')
df['session_end'] = pd.to_datetime(df['session_end'], format='%d.%m.%Y %H:%M')

print(df.dtypes)

user_id                   int64
region                      str
device                      str
channel                     str
session_start    datetime64[us]
session_end      datetime64[us]
time_session            float64
click_count             float64
buy_count               float64
price                   float64
age                       int64
dtype: object


## 6. Группировки и сводные таблицы 

### Задание 1. Группировка region и количество устройств (device)

In [25]:
task1_group = df.groupby(['region', 'device'])['user_id'].count()
task1_group

region         device 
Russia         Android     11
               Mac         12
               PC           6
               iPhone      23
United States  Android    174
               Mac        232
               PC          97
               iPhone     399
Name: user_id, dtype: int64

### Задание 2. Группировка device и количество рекламных источников каждого типа (channel)

In [26]:
task2_df = df.groupby(['device', 'channel']).size().reset_index(name='count')
task2_df = task2_df.sort_values(by='count', ascending=False).reset_index(drop=True)
task2_df

,device,channel,count
0,iPhone,organic,249
1,Mac,organic,174
2,Android,organic,115
3,iPhone,TipTop,78
4,PC,organic,72
5,iPhone,FaceBoom,65
6,Mac,FaceBoom,36
7,Android,FaceBoom,33
8,iPhone,MediaTornado,30
9,Android,TipTop,26


### Задание 3. Сводная таблица (pivot_table) — количество пользователей для каждого канала (channel)

In [27]:
task3_pivot = df.pivot_table(index='channel', values='user_id', aggfunc='count').rename(columns={'user_id': 'count'})
task3_pivot = task3_pivot.sort_values(by='count', ascending=False)
task3_pivot

,count
channel,
organic,610
FaceBoom,149
TipTop,142
MediaTornado,53


### Задание 4. Сводная таблица (pivot_table) — количество пользователей для каждого устройства (device — строки) и канала (channel — столбцы)

In [28]:
task4_pivot = df.pivot_table(index='device', columns='channel', values='user_id', aggfunc='count', fill_value=0)
task4_pivot = task4_pivot.sort_index(ascending=True)
task4_pivot

channel,FaceBoom,MediaTornado,TipTop,organic
device,,,,
Android,33,11,26,115
Mac,36,10,24,174
PC,15,2,14,72
iPhone,65,30,78,249


## Выводы по работе

В ходе выполнения лабораторной работы была проведена предварительная обработка набора данных `visits.csv` (954 записи), содержащего статистику сессий интернет-магазина.

**Основные результаты:**
1. Проведена проверка качества данных: устранен хвостовой пробел в имени первичного ключа (`user_id `), выявлены и заполнены модальными значениями пропуски в столбцах `region` ('United States') и `device` ('iPhone').
2. Устранены неявные дубликаты: синоним `USA` приведен к `United States`, а регистр названий устройств `MAC` и `IPHONE` стандартизирован к `Mac` и `iPhone`. Явных повторяющихся строк в выборке не зафиксировано.
3. Выполнено преобразование типов: строковые значения `session_start` и `session_end` переведены в формат `datetime64`.
4. На основе группировок и сводных таблиц варианта 6 установлено:
   - Подавляющая часть трафика относится к региону United States (902 визита из 954), а наиболее популярным устройством является iPhone (422 визита);
   - Ведущим источником привлечения пользователей выступает органический трафик (`organic` — 610 переходов), среди рекламных каналов равную отдачу имеют `FaceBoom` (149) и `TipTop` (142), наименьший объем генерирует `MediaTornado` (53);
   - Максимальная плотность визитов приходится на связку смартфона `iPhone` и канала `organic` (249 сессий).